cmd/multiFaAcceleration/multiFaAcceleration.go

```diff
@@ -0,0 +1,504 @@
+	b3         float64
+}
+
+//A set of all observed pairwise distances between the four species. While these numbers must be integers, we store them as float64 to avoid casting in other functions as a way to improve readability.
+type Distances struct {
+	D01 float64
+	D02 float64
+	D03 float64
+	D12 float64
+	D13 float64
+	D23 float64
+}
+
+//The set of branch lengths corresponding to a particular distance matrix.
+type BranchLengths struct {
+	B1 float64
+	B2 float64
+	B3 float64
+	B4 float64
+	B5 float64
+}
+
+//SubTree is a tree with three leaves and three branches, joined at the single internal node.
+type SubTree struct {
+	Dab float64
+	Dac float64
+	Dbc float64
+	va  float64
+	vb  float64
+	vc  float64
+}
+
+func multiFaAcceleration(s Settings) {
+	records := fasta.Read(s.InFile)
+	var searchSpace []bed.Bed
+	var referenceLength, i, j, threshold int
+	var bitArray []byte
+
+	//a couple of safety checks
+	if len(records) != 4 {
+		log.Fatalf("multiFaAcceleration accepts a multiFa file with 4 records, found %v.", len(records))
+	}
+	if len(records[1].Seq) != len(records[0].Seq) || len(records[2].Seq) != len(records[0].Seq) || len(records[3].Seq) != len(records[0].Seq) {
+		log.Fatalf("Error. All records must be of the same sequence length.")
+	}
+
+	referenceLength = fasta.AlnPosToRefPos(records[0], len(records[0].Seq)-1)
+
+	if s.SearchSpaceBed != "" {
+		//if we want to limit our search to a particular set of regions, we populate the bitArray with 1s instead of 0s at positions overlapping the input bed entries.
+		searchSpace = bed.Read(s.SearchSpaceBed)
+		bitArray = make([]byte, referenceLength)
+		for i = range searchSpace {
+			if searchSpace[i].Chrom == s.ChromName {
+				for j = searchSpace[i].ChromStart; j < searchSpace[i].ChromEnd; j++ {
+					bitArray[j] = 1
+				}
+			}
+		}
+		threshold = int(s.SearchSpaceProportion * float64(s.WindowSize)) //the minimum number of bases at which a window must overlap the search space in order to be considered a valid window.
+	}
+
+	var currDistances Distances
+	var distanceCache = make(map[Distances]BranchLengths)
+	var referenceCounter int = 0
+	var reachedEnd bool = false
+	var b1, b3 float64
+	var currCount int
+	var pass, containedInMap bool
+
+	//variables for normalization
+	var velSum, initialSum float64 = 0, 0
+	var branchCacheSlice = make([]BranchCache, 0)
+
+	for alignmentCounter := 0; reachedEnd == false && referenceCounter < referenceLength-s.WindowSize; alignmentCounter++ {
+		if s.Verbose && alignmentCounter%1000000 == 0 {
+			log.Printf("alignmentCounter: %v\n", alignmentCounter)
+		}
+		currCount, pass = thresholdCheckPasses(s, currCount, threshold, bitArray, referenceCounter) //first we see if we have a valid window.
+		if records[0].Seq[alignmentCounter] != dna.Gap {                                            //and if we are at a reference position.
+			if pass {
+				if s.UseSnpDistance {
+					reachedEnd = fourWaySnpDistances(records, alignmentCounter, s, &currDistances)
```

Comment:
Not a correctness problem, but this should be a target for any future optimization. Unless I am missing something, it seems like this recalculates distance from scratch for each window instead of just looking at the delta of the new bases after sliding down a base. I don't know how much of the computation is calculating the distances, but if it is significant, then the delta approach would be a massive performance boost.

Reply:
Dan has a good point if you find you want to speed this up.
Another minor thing you could think about if you need more speed would be passing in initial values for beginning the optimization.  I think the optimal values of the previous window will be a good guess at the optimal values of the current window.